molecule/structure_checker2.h

```diff
@@ -0,0 +1,150 @@
+/****************************************************************************
+ * Copyright (C) from 2009 to Present EPAM Systems.
+ *
+ * This file is part of Indigo toolkit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ***************************************************************************/
+
+#ifndef __structure_checker2__
+#define __structure_checker2__
+
+#include "base_c/defs.h"
+#include "base_cpp/exception.h"
+#include <string>
+#include <vector>
+#include <unordered_set>
+
+class IndigoObject;
+
+namespace indigo
+{
+
+    class BaseMolecule;
+
+    class DLLEXPORT StructureChecker2
```

Comment:
Is this class to supersede `StructureChecker`? Otherwise, please do not repeat nearly the same enums, moreover public, but reuse them (e.g. by inheritance or extracting to some common header...)

Reply:
Yes, the StructureChecker2.* is full replacement for StructureChecker.* and ReactionChecker.*

The functions "check2" in C and Python replace the corresponding functions "check"

After debugging, "check" instances should be removed and "check2" instances should be renamed to "check"

Reply:
OK. So finally (merged or squashed), we will be able to trace differences in-place (in structure_checker.h, in structure_checker.cpp, etc.)?

Reply:
Yes but this difference is up to 100%.
I've re-organized this "spaghetti" code so that each check is a separate independent (and unit-testable) function. See the functions referred in this array: 
static void (*check_type_checkers[])(Molecule&, const std::unordered_set<int>&, const std::unordered_set<int>&, StructureChecker2::CheckResult&) = {
    &check_load,         &check_valence,      &check_radical, &check_pseudoatom, &check_stereo,     &check_query,
    &check_overlap_atom, &check_overlap_bond, &check_rgroup,  &check_sgroup,     &check_tgroup,     &check_chirality,
    &check_chiral_flag,  &check_3d_coord,     &check_charge,  &check_salt,       &check_ambigous_h, &check_coord};

And joined structure_checker and reaction_checker with automatic detection of kind of checked item.

So the code is taken from the current checkers and python REST code but its fragments are combined by another way and git can't find them in prev version.

Reply:
>please do not break the history of the original file(s), i.e. do not use git rm
OK, I'll just copy-paste structure_checker2.* content to structure_checker.* and delete structure_checker2.* and reaction_checker.*

But this should be done after final approve only as a final cleanup. Now we can see, run and compare the both codes and this is helpful for testing & debugging.